client/transaction-pool/graph/src/validated_pool.rs

```diff
@@ -521,14 +531,55 @@ impl<B: ChainApi> ValidatedPool<B> {
 	pub fn status(&self) -> PoolStatus {
 		self.pool.read().status()
 	}
+
+	/// Notify all watchers that transactions in the block with hash have been finalized
+	pub async fn on_block_finalized(&self, block_hash: BlockHash<B>) -> Result<(), B::Error> {
```

Comment:
Can we be sure this method is not called concurrently? Since it's an async task it has a racy behavior that may cause the same work being done multiple times (if we receive two finalization events in a short succession).
While I think the notifications won't be sent twice (since the watchers are removed on finalisation), I'd still fix the method just to be safe. My proposal is to update `last_finalized_hash` right away (and switch to `Mutex` or some atomic cell type), i.e.:
```
let last_finalized = mem::replace(&mut self.last_finalized_hash.lock(), block_hash);
```